### ISSUE TITLE
Harvesting2.0 Architecture/Diagram

### ISSUE THREAD

**Issue by jbrown-xentity** [GSA/data.gov#4299]
## User Story

In order to show the harvesting2.0 diagram, data.gov team wants to build a true diagram of the harvesting2.0 paradigm.

## Acceptance Criteria

[ACs should be clearly demoable/verifiable whenever possible. Try specifying them using [BDD](https://en.wikipedia.org/wiki/Behavior-driven_development#Behavioral_specifications).]

- [ ] GIVEN someone wants to understand the harvesting2.0 architecture \
  WHEN the harvesting2.0 document is reviewed \
  THEN the architecture is understood \

## Background

https://github.com/GSA/data.gov/wiki/Harvesting-2.0

## Security Considerations ([required](https://nvd.nist.gov/800-53/Rev4/control/CM-4))

[comment]: # "Our SSP says 'The Data.gov team ensures security implications are considered as part of the agile requirements refinement process by including a section in the issue template used as a basis for new work.' so please don't remove this section without care."
None

## Sketch
We want the final architecture diagrams saved and loaded into the wiki. Ideally let's save the raw file somewhere (drive, github, etc) so if we need to make edits we're not going back to the drawing board...

- [Starting point mural board](https://app.mural.co/t/gsa6/m/gsa6/1680717254037/309427fcdba78e5e134c5f95350e9a6a8a670ab8?sender=ee431e2b-d904-4301-890c-6bc7c39b9a77)
- [Whiteboard pictures](https://drive.google.com/drive/u/2/folders/13a_3pg6gDoGVU_VMi8kNckCcECV3HDYY)

Tools we have used in the past for architecture, as possible places to start (could use any tool):
- Mural (see above)
- [Plant UML](https://github.com/GSA/datagov-compliance)

**Comment by nickumia-reisys** [comment]
First Diagram Created: [Harvest Job Lifecycle](https://drive.google.com/file/d/174fsTVgQCIZh0QOihR2bRZG0dpEgcMIj/view?usp=share_link)

**Comment by nickumia-reisys** [comment]
Second Diagram: [Verification + Validation of Harvester](https://drive.google.com/file/d/1CLdAs1aLNbeRN1cZcl24XurMSbrf1tEe/view?usp=share_link)
- We don't know if we will use this diagram, but I think it's a robust concept that we can use to ensure no duplicates and proper agency-to-data.gov linking

**Comment by nickumia-reisys** [comment]
Third Diagram: [Component Outline](https://drive.google.com/file/d/12sT1xxYWcM8_ysuHS35yy-Uqq_WNzUBI/view?usp=share_link)
- This needs more detail.  It was supposed to be high-level overview of the various aspects (I suppose of the controller...)
- I just put the logical inputs/outputs.
- For the abstract data inputs/outputs, we can add it to this diagram or make a more specific one for each operation/action.
- This definitely requires more feedback...

**Comment by nickumia-reisys** [comment]
Fourth (and final?) Diagram: [Controller Outline](https://drive.google.com/file/d/1jExSqgB5H_fkvdO85CGgQob0y8CNiX4V/view?usp=share_link)
- This is still a work in progress, but since it has an initial iteration, I figured I would post it.
- I abstracted technology from the logic (e.g. we are storing the "harvested catalog reference" in S3, but knowing it is in S3 doesn't help this diagram)
  - The output from `extract` isn't well-formed.  I think there's additional discussion that needs to happen around the structure and specific data types we'd like to use.
  - "harvested catalog reference" is meant to be a pointer as opposed to the actual data itself.
    - It may even be a pointer to a pointer. A reference to the list of filenames which are then pointers to the real data in S3.
- We probably need additional diagrams for each of the processes (`extract`, `validate`, `compare`, `transform` and `load`)
- I think this diagram misses the point of what the features of the `controller` are.  This diagram is for the hooks between the `controller` and the other services.  We just need to define the features and capabilities of the `controller itself`
  - i.e. how are `job_id`s tracked?  how are the steps triggers?  What is the flow logic within the controller itself?
